src/leo_cache_tran.erl

```diff
@@ -0,0 +1,198 @@
+    ets:new(?REPLYDB, [named_table, set, private]),
+    {ok, unused, ?TRAN_TIMEOUT}.
+
+handle_call(stop, _From, State) ->
+    {stop, normal, ok, State};
+
+%%handle_call({unregister, Tbl, Key}, From, State) ->
+%%    {Pid, _Ref} = From,
+%%    case ets:lookup(?REPLYDB, {Tbl, Key}) of
+%%        [{{Tbl, Key}, ReplyList}] ->
+%%            ?debugVal(ReplyList),
+%%            ReplyList2 = lists:filter(fun({Pid2, _Ref2}) ->
+%%                                              Pid2 =/= Pid
+%%                                      end, ReplyList),
+%%            ?debugVal(ReplyList2),
+%%            ets:insert(?REPLYDB, {{Tbl, Key}, ReplyList2});
+%%        _ ->
+%%            void
+%%    end,
+%%    {reply, ok, State, ?TRAN_TIMEOUT};
+
+handle_call({tran, Pid, Tbl, Key}, _From, State) ->
+    MonitorRef = erlang:monitor(process, Pid),
+    ets:insert(?PROCESSDB, {MonitorRef, {Tbl, Key}}), 
+    ets:insert_new(?REPLYDB, {{Tbl, Key}, []}),
+    {reply, ok, State, ?TRAN_TIMEOUT}; 
+
+handle_call({has_tran, Tbl, Key}, From, State) ->
+    case ets:lookup(?REPLYDB, {Tbl, Key}) of
+        [{{Tbl, Key}, ReplyList}] ->
+            ets:insert(?REPLYDB, {{Tbl, Key}, [From | ReplyList]}),
+            {noreply, State, ?TRAN_TIMEOUT};
+        _ ->
+            {reply, {ok, not_found}, State, ?TRAN_TIMEOUT}
+    end;
+
+handle_call({done_tran, Tbl, Key}, _From, State) ->
+    reply_all(Tbl, Key),
+    {reply, ok, State, ?TRAN_TIMEOUT}.
+
+handle_cast(_Msg, State) ->
+    {noreply, State, ?TRAN_TIMEOUT}.
+
+
+%% Function: handle_info(Info, State) -> {noreply, State}          |
+%%                                       {noreply, State, Timeout} |
+%%                                       {stop, Reason, State}
+%% Description: Handling all non call/cast messages
+%% handle_info({_Label, {_From, MRef}, get_modules}, State) ->
+%%     {noreply, State};
+handle_info({'DOWN', MonitorRef, _Type, _Pid,_Info}, State) ->
+    case ets:lookup(?PROCESSDB, MonitorRef) of
+        [{MonitorRef, {Tbl, Key}}] ->
+            reply_all(Tbl, Key);
+        _ ->
+            void
+    end,
+    ets:delete(?PROCESSDB, MonitorRef),
+    erlang:demonitor(MonitorRef),
+    {noreply, State, ?TRAN_TIMEOUT};
+handle_info(_Info, State) ->
+    {noreply, State, ?TRAN_TIMEOUT}.
+
+
+%% Function: terminate(Reason, State) -> void()
+%% Description: This function is called by a gen_server when it is about to
+%% terminate. It should be the opposite of Module:init/1 and do any necessary
+%% cleaning up. When it returns, the gen_server terminates with Reason.
+%% The return value is ignored.
+terminate(_Reason,_State) ->
+    ets:foldl(fun({_Key, ReplyList}, Acc) ->
+                      lists:foreach(fun(Target) ->
+                                            gen_server:reply(Target, {error, terminated})
+                                    end, ReplyList),
+                      Acc
+              end, [], ?REPLYDB),
+    ets:delete(?REPLYDB),
```

Comment:
No need this line because Erlang VM automatically delete all ets tables holden by the terminating process.